src/util/snapshot.cpp

```diff
@@ -0,0 +1,30 @@
+#include <faabric/util/memory.h>
+#include <faabric/util/snapshot.h>
+
+namespace faabric::util {
+
+std::vector<SnapshotDiff> SnapshotData::getDirtyPages()
+{
+    if (data == nullptr || size == 0) {
+        std::vector<faabric::util::SnapshotDiff> empty;
+        return empty;
+    }
+
+    // Get dirty pages
+    int nPages = getRequiredHostPages(size);
+    std::vector<bool> dirtyFlags = faabric::util::getDirtyPages(data, nPages);
+
+    // Convert to snapshot diffs
+    // TODO - reduce number of diffs by merging adjacent dirty pages
```

Comment:
Could we tackle the `TODO` in this PR? I feel it should not be very complex to implement. I was thinking of something like:
```cpp
auto next1 = std::find(dirtyFlags.begin(), dirtyFlags.end(), 1);
assert(next != dirtyFlags.end());
auto next0 = [&](){ return std::find(next1, dirtyFlags.end(), 0); };
while (next0() != dirtyFlags.end()) {
    // May have to +/- 1
    auto size = next0() - next1;
    uint32_t offset = next1 * faabric::util::HOST_PAGE_SIZE;
    diffs.emplace_back(offset, data + offset,  size * faabric::util::HOST_PAGE_SIZE;
    next1 = std::find(next0, dirtyFlags.end(), 1);
}
// handle the last chunk (i.e. when the vector ends with ones)
```
there's provably a nicer way, and the code above will be full of bugs, but just an idea.

Reply:
I think this may be premature optimisation as I'm not sure how often this would benefit (and will add yet more complexity to something that's already quite complicated). I'll revisit this TODO once I've done some benchmarking.